File: 1190. Reverse Substrings Between Each Pair of Parentheses.java
class Solution {
    public String reverseParentheses(String s) {
        Deque<Integer> stack = new ArrayDeque<>();
        Map<Integer, Integer> tele = new HashMap();
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c == '(') {
                stack.addLast(i);
            } else if (c == ')') {
                int prev = stack.pollLast();
                tele.put(i, prev);
                tele.put(prev, i);
            }
        }
        StringBuilder sb = new StringBuilder();
        for (
            int i = 0, direction = 1;
            i < s.length();
            i += direction
        ) {
            char c = s.charAt(i);
            if (c != '(' && c != ')') {
                sb.append(c);
                continue;
            }
            i = tele.get(i);
            direction = -direction;
        }
        return sb.toString();
    }
}

File: go/1190. Reverse Substrings Between Each Pair of Parentheses.go
func reverseParentheses(s string) string {
	stack := []int{}
	tele := make(map[int]int)
	for i := 0; i < len(s); i++ {
		if s[i] == '(' {
			stack = append(stack, i)
		} else if s[i] == ')' {
			prev := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			tele[i] = prev
			tele[prev] = i
		}
	}
	direction, i := 1, 0
    var result strings.Builder
	for i < len(s) {
		if s[i] != '(' && s[i] != ')' {
			result.WriteByte(s[i])
		} else {
			direction = -direction
			i = tele[i]
		}
		i += direction
	}
	return result.String()
}